node-sqlite3を使用して非同期処理を行う

nakamu-kazu222 opened the pull request:
- 実装すること
1.  node-sqlite3 をコールバックの形式で使う
2.  1 で使用したコールバック形式の関数をラップして Promise 化し、then, catch の形式で使う
3.  2 で作成した関数を async, await の形式で使う
  - さらに 1, 2, 3 の中で、それぞれ以下のことを行う
    - エラーなしのプログラムを書く
      - テーブルを作成する
      - レコードを追加し、自動採番された ID を標準出力に出力する
      - レコードを取得し、それを標準出力に出力する
      - テーブルを削除する
    - エラーありのプログラムを書く
      - テーブルを作成する
      - エラーを伴うレコードの追加を行い、発生したエラーのエラーメッセージを標準エラー出力に出力する
      - エラーを伴うレコードの取得を行い、発生したエラーのエラーメッセージを標準エラー出力に出力する
      - テーブルを削除する

nakamu-kazu222 committed:
node-sqlite3をインストールした

nakamu-kazu222 committed:
ESLint と Prettier でチェックできるようにした

nakamu-kazu222 committed:
node-sqlite3 をコールバックの形式で使い、エラーなし、エラーありのプログラムを作成した

nakamu-kazu222 committed:
Promiseを使い、エラーなし、エラーありのプログラムを作成した

nakamu-kazu222 committed:
async, awaitを使い、エラーなし、エラーありのプログラムを作成した

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+const db1 = new sqlite3.Database(":memory:");
+
+db1.run(
+  "CREATE TABLE IF NOT EXISTS book (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)",
```
この時点でテーブルが存在している可能性はないはずなので `IF NOT EXISTS` を付けても意味がない気がします。

nakamu-kazu222 replied:
この時点でテーブルが存在している可能性はないので、`IF NOT EXISTS`をつけないようにしました

cafedomancer replied:
他の箇所も同様に修正してください。

nakamu-kazu222 replied:
CREATE文のすべての`IF NOT EXISTS`を消去しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+          function (err, row) {
+            if (err) {
+              console.error("Error retrieving record:", err.message);
+            } else {
```
エラーなしのプログラムにはエラーハンドリングは不要です。あってもよいですが、SELECT の処理にはそれがあって、INSERT の処理にはないのは一貫性に欠けています。

nakamu-kazu222 replied:
エラーなしのプログラムのエラーハンドリングを消去しました

cafedomancer replied:
他のプログラムも同様に修正してください。

nakamu-kazu222 replied:
promise.js、async_await.jsのエラーなしのプログラムのエラーハンドリングを消去しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+              console.log("Retrieved record:", row);
+            }
+
+            timers.setTimeout(function () {
```
何も待つ必要はない気がするのですが、何を待っているんでしょうか？

nakamu-kazu222 replied:
次の処理を実行しても問題なく、待つ必要がないので消去しました

cafedomancer replied:
このコードを追加することになった経緯を確認したかったのですが、コードを読んだところでは分からなかったので何を待っているのかを確認していました。変更した結果を書く必要もあるのですが、何を待っているつもりだったのかをまず書いてほしかったです。また、他の同様な箇所についても必要なのかを改めて考えてみてもらいたいです。

nakamu-kazu222 replied:
申し訳ございません
回答など不足しておりました

SELECT文を実行し`console.log("Retrieved record:", row);`を出力した後にBookテーブルを削除した方がいいと思っておりました
テーブルを削除する前に、SELECT文の処理が終了していない場合を考えていました

しかし、待機しなくてもSELECT文の処理後に削除されるので、待機するコードは削除しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+            }
+
+            timers.setTimeout(function () {
+              db1.run("DROP TABLE IF EXISTS book", function () {
```
この時点でテーブルは必ず存在していないといけないので `IF EXISTS` は不要です。

nakamu-kazu222 replied:
テーブルが作成された後には必ずテーブルが存在するので、IF EXISTS を省略しました

cafedomancer replied:
他の箇所も同様に修正してください。また `IF EXISTS` はデフォルトの挙動ではないので、「省略」という言葉は適切ではないと思います。「省略」と書かれていると、書かなくても同じ挙動をするように見えますが、そうではないからです。

nakamu-kazu222 replied:
おっしゃる通りで、同じ挙動ではないので言葉の解釈が異なりますね

`IF EXISTS`があるDROP文を修正し、`IF EXISTS`を消去しました

cafedomancer commented on 03.asynchronous/package.json:
```diff
@@ -1,12 +1,16 @@
     "fix": "prettier --write . && eslint --fix .",
-    "lint": "prettier --check . && eslint ."
+    "lint": "prettier --check . && eslint .",
+    "format": "eslint '*.js' --fix && prettier --write '*.js'"
```
`"fix"` の内容とほぼ同じに見えるのですがこれを追加している理由はなんですか？

nakamu-kazu222 replied:
fixですべてのファイルに対して実行しているので、不要でした

cafedomancer replied:
不要だとは思いますが、追加している理由をまず確認しています。

nakamu-kazu222 replied:
すべてのファイルではなく、JSファイルに絞って実行すべきだと考えていました
ご指摘をいただいて学習し直して、`fix`だけでいいと思いました

cafedomancer replied:
> すべてのファイルではなく、JSファイルに絞って実行すべきだと考えていました

そうしたいのであれば、既存の `"fix"` を修正したほうがと思います。また、合わせて `"lint"` も修正しなければならないはずです。ただ、JavaScript のファイルに限定すべきかどうかは微妙な気がしました。`package.json` などのファイルの内容も変更されることがあり、場合によってはフォーマットに沿っていないことがあるからです。

nakamu-kazu222 replied:
> package.json などのファイルの内容も変更されることがあり、場合によってはフォーマットに沿っていないことがある

そうですね、`package.json`も含めたいので、すべてのファイルに対して実行するようにします

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,91 @@
+      }
+    });
+  });
+}
```
`promise.js` と関数の定義が重複しているので共通化してください。

nakamu-kazu222 replied:
重複している処理を別ファイルにて共通化しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+        db1.get(
+          "SELECT * FROM book WHERE id = ?",
+          [this.lastID],
+          function (err, row) {
```
関数の内部で `this` を使っていないのでアロー関数でよいはずです。

nakamu-kazu222 replied:
アロー関数に変換しました

cafedomancer replied:
同様の箇所が他にもあるので見直して修正してください。

nakamu-kazu222 replied:
callback.jsのエラーありの`function (err, row)`をアロー関数に変更しました

cafedomancer replied:
まだたくさん残っているような...。
https://github.com/nakamu-kazu222/js-practices/blob/2658ec604a5a0389dd33e116fef4520fa4fd6f73/03.asynchronous/callback.js#L8
https://github.com/nakamu-kazu222/js-practices/blob/2658ec604a5a0389dd33e116fef4520fa4fd6f73/03.asynchronous/callback.js#L18
https://github.com/nakamu-kazu222/js-practices/blob/2658ec604a5a0389dd33e116fef4520fa4fd6f73/03.asynchronous/callback.js#L19
https://github.com/nakamu-kazu222/js-practices/blob/2658ec604a5a0389dd33e116fef4520fa4fd6f73/03.asynchronous/callback.js#L35
https://github.com/nakamu-kazu222/js-practices/blob/2658ec604a5a0389dd33e116fef4520fa4fd6f73/03.asynchronous/callback.js#L39
https://github.com/nakamu-kazu222/js-practices/blob/2658ec604a5a0389dd33e116fef4520fa4fd6f73/03.asynchronous/callback.js#L53
https://github.com/nakamu-kazu222/js-practices/blob/ac35abd485a4f5be44bba7c5b208b7060d4aab46/03.asynchronous/run_get_query.js#L15

nakamu-kazu222 replied:
関数の内部で `this` を使っていないすべての`function ()`をアロー関数に変更しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+import sqlite3 from "sqlite3";
+import timers from "timers";
+
+const db1 = new sqlite3.Database(":memory:");
+
+db1.run(
+  "CREATE TABLE IF NOT EXISTS book (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)",
+  function () {
+    db1.run(
+      "INSERT INTO book (title) VALUES (?)",
+      ["Sample Title"],
+      function () {
+        console.log("Inserted record ID:", this.lastID);
+
+        db1.get(
+          "SELECT * FROM book WHERE id = ?",
+          [this.lastID],
+          function (err, row) {
+            if (err) {
+              console.error("Error retrieving record:", err.message);
+            } else {
+              console.log("Retrieved record:", row);
+            }
+
+            timers.setTimeout(function () {
+              db1.run("DROP TABLE IF EXISTS book", function () {
+                db1.close();
+
+                runErrorProgram();
+              });
+            }, 100);
+          }
+        );
+      }
+    );
+  }
+);
```
エラーありのプログラムを関数として定義するのであれば、エラーなしのプログラムも関数として定義するべきだと思います。

nakamu-kazu222 replied:
エラーなしのプログラムも関数として定義しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+            function (err, row) {
+              if (err) {
+                console.error("Error retrieving record:", err.message);
+              } else {
+                console.log("Retrieved record:", row);
+              }
```
SELECT には正常時の処理があって INSERT にはない理由が分かりません。どちらもありにする・どちらもなしにする、のどちらかに統一すべきだと思います。

nakamu-kazu222 replied:
INSERTでも正常時の処理を定義するようにしました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,72 @@
+
+            timers.setTimeout(function () {
+              db1.run("DROP TABLE IF EXISTS book", function () {
+                db1.close();
+
+                runErrorProgram();
```
close が完了してから次の処理を行ったほうがよいと思います。

nakamu-kazu222 replied:
close が完了してから次の処理を行うようにしました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,95 @@
+import sqlite3 from "sqlite3";
+import { setTimeout } from "timers/promises";
```
`callback.js` と import の形式が異なっています。どちらかのスタイルに統一したほうがよいと思います。

```
import timers from "timers";
```

nakamu-kazu222 replied:
`import { setTimeout } from "timers/promises";`のスタイルにしました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,95 @@
+}
+
+function runNoErrorProgram() {
+  return new Promise((resolve, reject) => {
+    const db1 = new sqlite3.Database(":memory:");
+
+    runQuery(
+      db1,
+      "CREATE TABLE IF NOT EXISTS book (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)"
+    )
+      .then(() =>
+        runQuery(db1, "INSERT INTO book (title) VALUES (?)", ["Sample Title"])
+      )
+      .then(() => getQuery(db1, "SELECT last_insert_rowid() as id"))
+      .then((lastID) => {
+        console.log("Inserted record ID:", lastID.id);
+        return getQuery(db1, "SELECT * FROM book WHERE id = ?", [lastID.id]);
+      })
+      .then((row) => {
+        console.log("Retrieved record:", row);
+        return setTimeout(100);
+      })
+      .then(() => runQuery(db1, "DROP TABLE IF EXISTS book"))
+      .then(() => {
+        db1.close();
+        resolve();
+      })
+      .catch((err) => {
+        console.error("Error in runNoErrorProgram:", err.message);
+        reject(err);
+      });
+  });
```
then や catch が promise を返すのだから、この処理を新しい promise で囲う必要はないはずです。

nakamu-kazu222 replied:
thenやcatch は新しいPromiseを返すため、promise で囲う必要はないので削除しました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,95 @@
+
+function runNoErrorProgram() {
+  return new Promise((resolve, reject) => {
+    const db1 = new sqlite3.Database(":memory:");
```
関数が分かれていて名前空間も分かれているのだから、`db1` や `db2` のような意味のない連番の名前をつけるのはやめましょう。

nakamu-kazu222 replied:
`noErrorDB`、`errorDB`と変更しました

cafedomancer replied:
連番をつけるのをやめましょうということです。今回のケースでは名前を具体的にする必要はないと思います。

nakamu-kazu222 replied:
関数が分かれていて名前空間も分かれているので同じ名前でも動作することを学習し、`db`という名前に修正しました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,95 @@
+      .then(() =>
+        runQuery(db1, "INSERT INTO book (title) VALUES (?)", ["Sample Title"])
+      )
+      .then(() => getQuery(db1, "SELECT last_insert_rowid() as id"))
```
SELECT するのではなく `runQuery` から得られる結果を使ってください。

nakamu-kazu222 replied:
runQuery から得られる結果を表示するようにしました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,95 @@
+      })
+      .then(() => runQuery(db1, "DROP TABLE IF EXISTS book"))
+      .then(() => {
+        db1.close();
```
close の処理も完了を待つべきだと思います。

nakamu-kazu222 replied:
closeの処理の完了を待機するようにしました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,91 @@
+async function runErrorProgram() {
+  const db2 = new sqlite3.Database(":memory:");
+
+  try {
+    await runQuery(
+      db2,
+      "CREATE TABLE IF NOT EXISTS book (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)"
+    );
+
+    await runQuery(db2, "INSERT INTO memo (title) VALUES (?)", [
+      "Sample Title",
+    ]);
+  } catch (err) {
```
例外が送出される想定の処理のみを `try ... catch` で囲んでください。

nakamu-kazu222 replied:
CREATE文とINSERT文の例外処理を分けました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,91 @@
+    await runQuery(db2, "INSERT INTO memo (title) VALUES (?)", [
+      "Sample Title",
+    ]);
+  } catch (err) {
+    console.error("Error inserting record:", err.message);
+  }
```
現状のコードだとすべての例外が捕捉されてしまいます。捕捉したい例外のみを捕捉するようにし、それ以外の例外は再送出するようにしてください。

nakamu-kazu222 replied:
INSERT文の例外のみ捕捉するようにしました

nakamu-kazu222 committed:
この時点でテーブルが存在している可能性はないはずなのでIF NOT EXISTSを消した

nakamu-kazu222 committed:
データベースの操作が非同期であるため、コールバック関数内で次の処理を実行しても問題ないので消去した

nakamu-kazu222 committed:
テーブルが作成された後には必ずテーブルが存在するので、IF EXISTS を省略した

nakamu-kazu222 committed:
fixですべてのファイルに対して実行するので削除した

nakamu-kazu222 committed:
関数の定義が重複しているので共通化した

nakamu-kazu222 committed:
エラーなしのプログラムも関数化した

nakamu-kazu222 committed:
エラーなしのプログラムにはエラーハンドリングは不要なので省略した

関数の内部で this を使っていないのでアロー関数に変更した

nakamu-kazu222 committed:
INSERTで正常時の処理を定義するようにした

nakamu-kazu222 committed:
close が完了してから次の処理を行うようにした

nakamu-kazu222 committed:
thenやcatch は新しいPromiseを返すため、promise で囲う必要はないので削除した

nakamu-kazu222 committed:
変数名を変更した

nakamu-kazu222 committed:
SELECT ではなく runQuery から得られる結果を表示した

nakamu-kazu222 committed:
closeの処理の完了を待機するようにした

nakamu-kazu222 committed:
promise.jsと同様にINSERT文から得られる結果を取得した

nakamu-kazu222 committed:
promise.jsと同様にcloseの処理の完了を待機した

nakamu-kazu222 committed:
CREATE文とINSERT文の例外処理を分けた

nakamu-kazu222 committed:
テーブルが存在している可能性はないはずなのでIF NOT EXISTSを消した

nakamu-kazu222 committed:
エラーなしのプログラムのエラーハンドリングを消去した

nakamu-kazu222 committed:
待機しなくてもSELECT文の処理後にテーブルが削除されるので、待機するコードは削除した

nakamu-kazu222 committed:
テーブルが作成された後には必ずテーブルが存在するので、IF EXISTSを削除した

nakamu-kazu222 committed:
関数の内部で this を使っていないのでアロー関数に変更した

nakamu-kazu222 committed:
関数が分かれていて名前空間も分かれているので同じ名前でも動作するので、名前を統一した

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,63 @@
+import sqlite3 from "sqlite3";
+
+const db = new sqlite3.Database(":memory:");
```
これは `runNoErrorProgram()` の中でしか使われないのだから `runNoErrorProgram()` の中で定義しておくべきだと思います。

nakamu-kazu222 replied:
runNoErrorProgram() の中に定義しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,63 @@
+          if (err) {
+            console.error("Error inserting record:", err.message);
+          } else {
+            console.log("Record inserted successfully");
```
エラーなしのプログラムと出力内容が異なっています。

nakamu-kazu222 replied:
エラーなしのプログラムと出力内容が同じように修正しました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,61 @@
+      runQuery(db, "INSERT INTO book (title) VALUES (?)", ["Sample Title"])
+    )
+    .then((result) => {
+      const lastID = result.lastID;
```
変数に入れなくてよいと思います。17,18 行目でそれぞれ `result.lastID` を使うようにしても計算量には大きな影響がないからです。`async_await.js` についても同様です。

nakamu-kazu222 replied:
変数に格納しないで、`result.lastID`を直接使用するようにしました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,61 @@
+      return getQuery(db, "SELECT * FROM book WHERE id = ?", [lastID]);
+    })
+    .then((row) => {
+      console.log("Retrieved record:", row);
+    })
+    .then(() => runQuery(db, "DROP TABLE book"))
```
`console.log()` は非同期処理を行う関数ではないので、その完了を `then` で待つ必要はないです。`console.error()` についても同様なので合わせて修正してください。

nakamu-kazu222 replied:
`console.log()`の完了を待たずに`runQuery(db, "DROP TABLE book")`を実行するようにしました

cafedomancer replied:
ちょっと細かいですが、`console.log()` の完了は待っているというか、実行が完了しないと次の処理に遷移しないので、「完了を待たずに」ではなく「`then` を使って待たずに」という表現が正しいのでは、と思います。

nakamu-kazu222 replied:
> 「then を使って待たずに」という表現が正しい

おっしゃる通りで、実行が完了しないと次の処理に遷移しないので上記の表現でした
ご指摘いただきありがとうございます

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,61 @@
+    })
+    .then(() => runQuery(db, "DROP TABLE book"))
+    .then(() => {
+      db.close();
+      return setTimeout(100);
```
せっかく promise の課題をやっているのだから、close も promise の形式で待てるようにしたほうがいいんじゃないですかね...？ `async_await.js` についても同様です。

nakamu-kazu222 replied:
promise の形式で待てるようにしました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,61 @@
+    .then(() => {
+      db.close();
+    })
+    .catch((err) => {
+      console.error("Error in runErrorProgram:", err.message);
+    });
```
コードを読んだところでは分からなかったのですが、何を catch する想定ですか？

nakamu-kazu222 replied:
`db.close()`でエラーが出た場合にcatch する想定でした
しかし、`db.close()`は非同期であるので、このままだとできないことに気づきました

ただ、`db.close()`がエラーを発生させる可能性が低いため、catchは消去したいと思います

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,61 @@
+
+runNoErrorProgram()
+  .then(() => runErrorProgram())
+  .catch((err) => console.error("Error:", err.message));
```
コードを読んだところでは分からなかったのですが、何を catch する想定ですか？

nakamu-kazu222 replied:
runNoErrorProgram および runErrorProgram でエラーが起きた時にcatchする想定でした

しかし、runNoErrorProgram および runErrorProgram の関数内でcatchがあるので消去します

cafedomancer commented on 03.asynchronous/run_get_query.js:
ファイル名が具体的すぎるのでもう少し抽象的な名前をつけたほうがよさそうです。

nakamu-kazu222 replied:
queryを実行するので、query.jsという抽象的な名前にしました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,66 @@
+async function runNoErrorProgram() {
+  const db = new sqlite3.Database(":memory:");
+
+  try {
+    await runQuery(
+      db,
+      "CREATE TABLE book (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)"
+    );
+
+    const result = await runQuery(db, "INSERT INTO book (title) VALUES (?)", [
+      "Sample Title",
+    ]);
+
+    const lastID = result.lastID;
+    console.log("Inserted record ID:", lastID);
+
+    const row = await getQuery(db, "SELECT * FROM book WHERE id = ?", [lastID]);
+    console.log("Retrieved record:", row);
+
+    await runQuery(db, "DROP TABLE book");
+  } finally {
+    db.close();
+    await setTimeout(100);
+  }
```
try で囲っている理由が分からなかったのですが、そうしているのはなぜですか？

nakamu-kazu222 replied:
修正を忘れました

catchがないので、try、finallyは消去しました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,66 @@
+  await runErrorProgram();
+}
+
+main().catch((err) => console.error("Error:", err.message));
```
コードを読んだところでは分からなかったのですが、何を catch する想定ですか？

nakamu-kazu222 replied:
runNoErrorProgram および runErrorProgram でエラーが起きた時にcatchする想定でした

しかし、runNoErrorProgram および runErrorProgram の関数内でcatchがあるので消去します

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,66 @@
+    }
+  } catch (err) {
+    console.error("Error creating table:", err.message);
+  }
```
`CREATE TABLE` はエラーを起こさない想定なのでエラーハンドリングはなくてもよいです。あってもよいですが、`try ... catch` をネストする必要はないのと、`DROP TABLE` にもエラーハンドリングがないと対称性を欠くことになると思います。

nakamu-kazu222 replied:
`CREATE TABLE`のエラーハンドリングを消去しました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,66 @@
+      await runQuery(db, "INSERT INTO memo (title) VALUES (?)", [
+        "Sample Title",
+      ]);
+    } catch (err) {
+      console.error("Error inserting record:", err.message);
+    }
```
`try ... catch` では例外として送出されたものはどんなものでも捕捉されてしまうので、捕捉したい例外のみを捕捉するようにしてそれ以外の例外は再送出するようにしてください。他の `try ... catch` についても同様です。

nakamu-kazu222 replied:
SQLのみ捕捉するようにしました

cafedomancer replied:
"SQL" は捕捉されないと思うので言葉遣いが正しくなさそうです... 👀

nakamu-kazu222 replied:
"SQLエラー"を捕捉するという表現で正しいでしょうか？

cafedomancer replied:
「SQL に関する例外を捕捉する」じゃないですかね 🤔

nakamu-kazu222 replied:
> SQL に関する例外を捕捉する

なるほどです
理解しました

nakamu-kazu222 committed:
関数の内部で this を使っていないすべてのfunction ()をアロー関数に変更した

nakamu-kazu222 committed:
runNoErrorProgram() の中でしか使われないので、中に定義した

nakamu-kazu222 committed:
エラーなしのプログラムと出力内容が同じように修正した

nakamu-kazu222 committed:
変数に入れなくても計算量には大きな影響がないので、修正した

nakamu-kazu222 committed:
正常時の処理を追加した

エラーなしのプログラムと同じ出力結果にした

nakamu-kazu222 committed:
console.logは非同期処理を行う関数ではなく、その完了をthenで待つ必要はないので修正した

nakamu-kazu222 committed:
promise の形式で待てるようにした

nakamu-kazu222 committed:
catchする必要がないので消去した

db.close()は非同期なので、catchできない
また、db.close()がエラーを発生させる可能性が低いため、catchは消去した

nakamu-kazu222 committed:
関数内でcatchがあるので消去した

nakamu-kazu222 committed:
ファイル名が具体的すぎるのでもう少し抽象的な名前をつけた

nakamu-kazu222 committed:
catchがないので、try、finallyは消去した

nakamu-kazu222 committed:
関数内でcatchがあるので消去した

nakamu-kazu222 committed:
CREATE TABLE はエラーを起こさない想定なのでエラーハンドリングを消去した

nakamu-kazu222 committed:
SQLのみ捕捉するようにした

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,59 @@
+      return runQuery(db, "DROP TABLE book");
+    })
+    .then(() => {
+      return setTimeout(() => {
+        db.close();
+      }, 100);
```
https://github.com/nakamu-kazu222/js-practices/pull/4#discussion_r1413934724 の続きになりますが、`close()` を promise の形式で使えるようにする必要がありますよ。つまり、`setTimeout()` を使うなということです。`async_await.js` についても同様です。また、関数本文が  return 1 行しかないときは簡潔文体を使ってください。

nakamu-kazu222 replied:
`.then(() => db.close());` に修正しました
`setTimeout()`がなくてもcloseの処理が終わり次第、runErrorProgram()の処理が始まるということですね

cafedomancer replied:
> `setTimeout()`がなくてもcloseの処理が終わり次第、runErrorProgram()の処理が始まるということですね

違います。https://github.com/nakamu-kazu222/js-practices/pull/4#discussion_r1416652059 のコメントと関連しますが、`db.close()` はコールバックベースで非同期処理を行う関数なので、promise の形式で待てるように自分で書き換えないと then の形式ではその完了は待てません。これだと void で resolve された promise が返るだけです。

```js
.then(() => db.close());
```

nakamu-kazu222 replied:
> db.close() はコールバックベースで非同期処理を行う関数なので、promise の形式で待てるように自分で書き換えないと then の形式ではその完了は待てません

おっしゃっている内容を理解しました

対応として、db.close() のコールバックをPromiseに変換して、resolve と reject を使用して非同期操作の結果を Promise に伝えるようにしました

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,59 @@
+    })
+    .catch((err) => {
+      console.error("Error retrieving record:", err.message);
+      return runQuery(db, "DROP TABLE book");
```
これだと promise が reject された場合にしか `DROP TABLE` が実行されないです。それはそれで問題ないのですが、43 行目の処理はそうなっていないので対称性がない点が気になっています。どちらも、reject された場合にのみ後続の処理を実行するようにするか、reject されたかどうかによらずに後続の処理を実行するようにするか、のどちらかにしてください。

nakamu-kazu222 replied:
43 行目の処理と揃えるために、rejectされたかどうかによらずに後続の処理を実行するようにしました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,70 @@
+  console.log("Retrieved record:", row);
+
+  await runQuery(db, "DROP TABLE book");
+  await db.close();
```
`db.close()` は promise を返す関数ではないので await では完了を待てません。

nakamu-kazu222 replied:
awaitをはずしました

closeが終わるまで待機したいと考えていましたが、
`await runNoErrorProgram();`でcloseが終わるまで待機しているので問題ないと思い、awaitを外しました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,70 @@
+    ]);
+    console.log("Inserted record ID:", result.lastID);
+  } catch (err) {
+    if (err.message.includes("no such table: memo")) {
```
このコードだと `err` にプロパティアクセスできないような値 (`null` や `undefined` など) が格納されているときにエラーが起きてしまいます。また、`no such table: memo` 以外のメッセージで例外が送出されるようにようになったときにも正しく動かなくなってしまうと思います。JavaScript の例外処理について調べて、もう少しうまくやる方法を考えてみてください。

nakamu-kazu222 replied:
`err instanceof Error`を使用するように修正しました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,70 @@
+    } else {
+      throw err;
+    }
+  } finally {
+    await runQuery(db, "DROP TABLE book");
+    await db.close();
+  }
```
この `try ... catch` で必ず実行したい処理というわけではなく、この関数の処理の最後の処理として行いたいものだと思うので、この処理は finally に入れなくてもよいと思います。

nakamu-kazu222 replied:
finally の外に`db.close();`を移しました

cafedomancer commented on 03.asynchronous/async_await.js:
```diff
@@ -0,0 +1,70 @@
+  }
+}
+
+async function main() {
+  await runNoErrorProgram();
+  await runErrorProgram();
+}
```
`main` 関数を定義したいのならこの書き方でもよいですが、そうでないのなら Top-level await を使って書いてもよいです。

nakamu-kazu222 replied:
Top-level await を使って書くように修正しました

nakamu-kazu222 committed:
setTimeout()がなくてもcloseの処理が終わり次第、次の処理が始まるので消去した

nakamu-kazu222 committed:
SELECTの処理と揃えるために、rejectされたかどうかによらずに後続の処理を実行するようにした

nakamu-kazu222 committed:
awaitを消去した

db.close() は promise を返す関数ではないので await では完了を待てない
await runNoErrorProgram();でcloseが終わるまで待機しているので問題ない

nakamu-kazu222 committed:
err instanceof Errorを使用するように修正した

err にプロパティアクセスできないような値 (null や undefined など) が格納されているときにエラーが起きてしまうので修正した

nakamu-kazu222 committed:
try ... catch で必ず実行したい処理というわけではないので、finallyの外に移した

nakamu-kazu222 committed:
Top-level await でも問題ないので修正した

nakamu-kazu222 committed:
db.close()をpromise の形式で待てるように自分で書き換えた

db.close() のコールバックをPromiseに変換して、resolve と reject を使用して非同期操作の結果を Promise に伝えるようにした

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,72 @@
+      return runQuery(db, "DROP TABLE book");
+    })
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        db.close((err) => {
+          if (err) {
+            reject(err);
+          } else {
+            resolve();
+          }
+        });
+      });
```
内容自体はよいのですが、エラーありのプログラムと定義が重複しているので共通化してください。また、async/await のプログラムも同様に修正してください。

nakamu-kazu222 replied:
close処理を共通化しました

nakamu-kazu222 committed:
close処理を共通化した

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,64 @@
+import sqlite3 from "sqlite3";
+import { runQuery, getQuery } from "./query.js";
+
+function closeDatabase(db) {
+  return new Promise((resolve, reject) => {
+    db.close((err) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve();
+      }
+    });
+  });
+}
```
`async_await.js` と定義がまだ重複していませんか...？

nakamu-kazu222 replied:
共通の定義を別ファイルに定義しました

nakamu-kazu222 committed:
共通の定義をファイルに切り出した

cafedomancer commented on 03.asynchronous/close_database.js:
このファイルの内容も node-sqlite3 の関数にまつわるものなので、現 `query.js` の内容とまとめたほうがいいんじゃないですかね...？

nakamu-kazu222 replied:
おっしゃる通りで、node-sqlite3 の関数にまつわるものなので統合すべきですね
見落としていました

query.jsに統合しました

nakamu-kazu222 committed:
closeDatabase(db)はnode-sqlite3 の関数にまつわるものなので、統合した

cafedomancer commented on 03.asynchronous/query.js:
close の処理は query ではないのでファイル名も見直す必要がありそうです。

nakamu-kazu222 replied:
抽象的な命名にするため、`sqlite.js`にしました

nakamu-kazu222 committed:
close の処理は query ではないのでファイル名も見直した
